Snowflake.Data/Core/SFChunkDownloader.cs

```diff
@@ -42,94 +41,70 @@ class SFChunkDownloader
+            _downloadTasks = new BlockingCollection<Task<SFResultChunk>>(prefetchSlot);
 
-            if (currentChunk.downloadState == DownloadState.SUCCESS)
+            Task.Run(() =>
             {
-                nextChunkToConsumeIndex++;
-                return currentChunk; 
-            }
-            else
-            {
-                // wait until donwload finish
-                lock(currentChunk.syncPrimitive)
+                foreach (var c in chunks)
                 {
-                    while(currentChunk.downloadState == DownloadState.IN_PROGRESS 
-                        || currentChunk.downloadState == DownloadState.NOT_STARTED)
+                    _downloadTasks.Add(DownloadChunkAsync(new DownloadContext()
                     {
-                        Monitor.Wait(currentChunk.syncPrimitive);
-                    }
+                        chunk = c,
+                        chunkIndex = nextChunkToDownloadIndex,
+                        qrmk = this.qrmk,
+                        chunkHeaders = this.chunkHeaders
+                    }));
                 }
-                nextChunkToConsumeIndex++;
-                return currentChunk;
-            }
-        } 
-        static void downloadChunkCallBack(Object context)
+
+                _downloadTasks.CompleteAdding();
+            });
+        }
+        
+        public SFResultChunk GetNextChunk()
+        {
+            return _downloadTasks.IsCompleted ? null : _downloadTasks.Take().Result;
+        }
+        
+        private async Task<SFResultChunk> DownloadChunkAsync(DownloadContext downloadContext)
         {
-            DownloadContext downloadContext = (DownloadContext)context;
             SFResultChunk chunk = downloadContext.chunk;
 
-            // change download status
-            Monitor.Enter(chunk.syncPrimitive);
-            try
-            {
-                chunk.downloadState = DownloadState.IN_PROGRESS;
-            }
-            finally
-            {
-                Monitor.Exit(chunk.syncPrimitive);
-            }
+            chunk.downloadState = DownloadState.IN_PROGRESS;
 
             S3DownloadRequest downloadRequest = new S3DownloadRequest()
             {
                 uri = new UriBuilder(chunk.url).Uri,
                 qrmk = downloadContext.qrmk,
                 // s3 download request timeout to one hour
-                timeout = 60 * 60,
-                httpRequestTimeout = 16000,
+                timeout = TimeSpan.FromHours(1),
+                httpRequestTimeout = TimeSpan.FromSeconds(16),
                 chunkHeaders = downloadContext.chunkHeaders
             };
 
-            HttpResponseMessage httpResponse = restRequest.get(downloadRequest);
+            var httpResponse = await restRequest.GetAsync(downloadRequest);
             Stream stream = httpResponse.Content.ReadAsStreamAsync().Result;
             IEnumerable<string> encoding;
+            //TODO this shouldn't be required.

```

Comment:
Can you explain why this should not be required?

Reply:
HttpClient supports Automatic Decompression of responses.

See: https://msdn.microsoft.com/en-us/library/system.net.http.httpclienthandler.automaticdecompression(v=vs.118).aspx

Need to confirm if it is enabled by default of if it needs to be requested. I did not get around to testing that yet.